aten/src/ATen/core/IList.h

```diff
@@ -0,0 +1,512 @@
+ * given that the container for `IListTag::Boxed` is `c10::List`, this
+ * iterator will wrap a `c10::List::const_iterator`.
+ */
+template <typename T>
+class IListIterator : public std::iterator<std::bidirectional_iterator_tag, T> {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...)                  \
+  friend class detail::IListTagImpl<IListTag::TAG, T>; \
+  friend class detail::IListTagImplBase<               \
+      IListTag::TAG,                                   \
+      T,                                               \
+      typename detail::IListTagImpl<IListTag::TAG, T>::elem_type>;
+  TORCH_ILIST_FORALL_TAGS(DEFINE_FRIEND_CLASS)
+#undef DEFINE_FRIEND_CLASS
+
+ public:
+  using unboxed_iterator_type = typename detail::
+      IListTagImpl<IListTag::Unboxed, T>::list_type::const_iterator;
+  using boxed_iterator_type = typename detail::
+      IListTagImpl<IListTag::Boxed, T>::list_type::const_iterator;
+
+  IListIterator() : tag_(IListTag::None) {}
+
+  IListIterator(boxed_iterator_type boxed) : tag_(IListTag::Boxed) {
+    payload_.boxed_iterator = boxed;
+  }
+
+  IListIterator(unboxed_iterator_type unboxed) : tag_(IListTag::Unboxed) {
+    payload_.unboxed_iterator = unboxed;
+  }
+
+  detail::IListConstRef<T> operator*() const {
+    TORCH_ILIST_UNWRAP(tag_, { return ImplT::iterator_get(this_); });
+  }
+
+  IListIterator& operator++() {
+    TORCH_ILIST_UNWRAP(tag_, { ++this_; });
+    return *this;
+  }
+
+  IListIterator operator++(int) {
+    auto old = *this;
+    TORCH_ILIST_UNWRAP(tag_, { ++this_; });
+    return old;
+  }
+
+  IListIterator& operator--() {
+    TORCH_ILIST_UNWRAP(tag_, { --this_; });
+    return *this;
+  }
+
+  IListIterator operator--(int) {
+    auto old = *this;
+    TORCH_ILIST_UNWRAP(tag_, { --this_; });
+    return old;
+  }
+
+  bool operator==(const IListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ILIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const IListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  union Payload {
+    boxed_iterator_type boxed_iterator;
+    unboxed_iterator_type unboxed_iterator;
+    void* _init_ptr;
+    Payload() : _init_ptr(nullptr) {}
+    ~Payload() {}
+  };
+
+  Payload payload_;
+  IListTag tag_;
+};
+
+/*
+ * Wrapper around different API containers (e.g. boxed and unboxed).
+ *
+ * What is it?
+ * ===========
+ * It is a tagged union of both boxed and unboxed API containers.
+ * Working implementations:
+ *
+ * - `IList<at::Tensor>`
+ * - `IList<at::OptionalTensorRef>`
+ *
+ * Note that `ITensorList` is a view type. Meaning that it won't own the
+ * tensors it holds. If you need it to last longer, make sure that there is
+ * actually a non-temporary list of tensors (e.g. `vector<Tensor>`) that owns
+ * them and outlives the `ITensorList` instance.
```

Comment:
It's probably better to just say that it is intended to be used only for argument parameters (where temporaries are guaranteed to live long enough because they will live for the entirety of the function call)